### PR TITLE
Fix comp to llvm memref conversion

### PR DESCRIPTION
**Pull request by kdobros**
Default function signature conversion also converts memrefs in function
signature which results in lowering not expected by runtime.
Proposed solution is to convert only `comp` dialect types here.

### CODE REVIEW

File: pmlc/conversion/comp_to_llvm/common.cc [hunk 1]
```diff
@@ -112,9 +114,57 @@ getManglingTypes(mlir::MLIRContext *context) {
       mlir::FloatType::getF32(context),    mlir::FloatType::getF64(context)};
 }
 
+namespace {
+struct ConvertFuncSignature : mlir::OpConversionPattern<mlir::FuncOp> {
+  ConvertFuncSignature(mlir::TypeConverter &typeConverter,
+                       mlir::MLIRContext *context)
+      : mlir::OpConversionPattern<mlir::FuncOp>(typeConverter, context) {}
```
Comment (earhart) [inline pmlc/conversion/comp_to_llvm/common.cc:121]: You can use `using mlir::OpConversionPattern<mlir::FuncOp>::OpConversionPattern;` to pick up the constructors directly.

File: pmlc/conversion/comp_to_llvm/common.cc [hunk 2]
```diff
@@ -112,9 +114,57 @@ getManglingTypes(mlir::MLIRContext *context) {
       mlir::FloatType::getF32(context),    mlir::FloatType::getF64(context)};
 }
 
+namespace {
+struct ConvertFuncSignature : mlir::OpConversionPattern<mlir::FuncOp> {
```
Comment (earhart) [inline pmlc/conversion/comp_to_llvm/common.cc:118]: This struct should be `final`

File: pmlc/conversion/comp_to_llvm/common.cc [hunk 3]
```diff
@@ -112,9 +114,57 @@ getManglingTypes(mlir::MLIRContext *context) {
...
+  ConvertFuncSignature(mlir::TypeConverter &typeConverter,
+                       mlir::MLIRContext *context)
+      : mlir::OpConversionPattern<mlir::FuncOp>(typeConverter, context) {}
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::FuncOp op, mlir::ArrayRef<mlir::Value> operands,
+                  mlir::ConversionPatternRewriter &rewriter) const override {
```
Comment (earhart) [inline pmlc/conversion/comp_to_llvm/common.cc:125]: This method should be `final`

**Review by earhart — CHANGES_REQUESTED** [review]
Is there any reason not to use `mlir::populateFuncOpTypeConversionPattern(patterns, context, typeConverter);` for this, using a type converter that returns its input for any non-`comp` type?  (And then you can use that for the dynamic legality check, too -- something like,

      target.addDynamicallyLegalOp<mlir::FuncOp>([&](mlir::FuncOp op) -> bool {
        return typeConverter.isSignatureLegal(op.getType());
      });

)

File: pmlc/conversion/comp_to_llvm/common.cc (hunk 2 above)
Comment (kdobros) [reply pmlc/conversion/comp_to_llvm/common.cc]: Done

File: pmlc/conversion/comp_to_llvm/common.cc (hunk 3 above)
Comment (kdobros) [reply pmlc/conversion/comp_to_llvm/common.cc]: Done

File: pmlc/conversion/comp_to_llvm/common.cc (hunk 1 above)
Comment (kdobros) [reply pmlc/conversion/comp_to_llvm/common.cc:121]: This pattern requires type converter, so I don't want to also pick up constructor that sets it to nullptr.

**Comment by kdobros** [comment]
> Is there any reason not to use `mlir::populateFuncOpTypeConversionPattern(patterns, context, typeConverter);` for this, using a type converter that returns its input for any non-`comp` type? (And then you can use that for the dynamic legality check, too -- something like,
> 
> ```
>   target.addDynamicallyLegalOp<mlir::FuncOp>([&](mlir::FuncOp op) -> bool {
>     return typeConverter.isSignatureLegal(op.getType());
>   });
> ```
> 
> )

That's a good point, I somehow rejected it after seeing two type converters. I changed it and now it looks much cleaner, there is no duplication of such simple functionality as converting function signature.

**Review by earhart — APPROVED** [review]

**Pull request merged**